src/grt/grt-rtiis.adb

```diff
@@ -0,0 +1,350 @@
+                  --Layout_Addr := Bt.Layout;
+                  null;
+               end if;
+            end;
+         when Ghdl_Rtik_Subtype_Record =>
+            declare
+               St : constant Ghdl_Rtin_Subtype_Composite_Acc :=
+                 To_Ghdl_Rtin_Subtype_Composite_Acc (Rti);
+            begin
+               Layout_Addr := Loc_To_Addr (St.Common.Depth, St.Layout, Ctxt);
+            end;
+         when others =>
+            Internal_Error("to_ghdl_rtii");
+      end case;
+      return Layout_Addr;
+   end Get_Type_Layout;
+
+   function Get_Array_Nbr_Children(Rtii : Ghdl_Rtii) return Ghdl_Index_Type is
+      Kind : Ghdl_Rtik;
+      Base_Array : Ghdl_Rtin_Type_Array_Acc;
+      Layout_Addr : Address;
+      Bounds_Addr : Address;
+      Length : Ghdl_Index_Type;
+      Idx_Rti : Ghdl_Rti_Access;
+      Rng : Ghdl_Range_Ptr;
+   begin
+      Kind := Rtii.Typ.Rti.Kind;
+      case Kind is
+         when Ghdl_Rtik_Type_Array =>
+            Base_Array := To_Ghdl_Rtin_Type_Array_Acc(Rtii.Typ.Rti);
+            Layout_Addr := Rtii.Typ.Binding_Layout_Addr;
+         when Ghdl_Rtik_Subtype_Array =>
+            Base_Array := To_Ghdl_Rtin_Type_Array_Acc(
+               To_Ghdl_Rtin_Subtype_Composite_Acc(Rtii.Typ.Rti).Basetype);
+            Layout_Addr := Rtii.Typ.Layout_Addr;
+         when others =>
+            Internal_Error("get_array_element_type");
+      end case;
+      Bounds_Addr := Layout_Addr + Ghdl_Index_Type'(
+         Ghdl_Indexes_Type'Size / 8);
+      -- Get the length of this dimension so we can check whether
+      -- the index is valid.
+      Idx_Rti := Get_Base_Type (Base_Array.Indexes (Rtii.Typ.Dim));
+      Extract_Range (Bounds_Addr, Idx_Rti, Rng);
+      Length := Range_To_Length (Rng, Idx_Rti);
+      return Length;
+   end Get_Array_Nbr_Children;
+
+   function Get_Array_Child(Rtii : Ghdl_Rtii; Index : Ghdl_Index_Type)
+                           return Ghdl_Rtii is
+      Kind : Ghdl_Rtik;
+      Base_Array : Ghdl_Rtin_Type_Array_Acc;
+      El_Type : Ghdl_Rtii_Type;
+      Layout_Addr : Address;
+      Binding_Layout_Addr : Address;
+      Bounds_Addr : Address;
+      Child : Ghdl_Rtii;
+      Length : Ghdl_Index_Type;
+      Idx_Rti : Ghdl_Rti_Access;
+      Rng : Ghdl_Range_Ptr;
+      El_Layout_Addr : Address;
+      El_In_Step : Ghdl_Index_Type;
+      Child_Addr : Address;
+      Sizes : Ghdl_Indexes_Ptr;
+      El_Size : Ghdl_Index_Type;
+   begin
+      Kind := Rtii.Typ.Rti.Kind;
+      case Kind is
+         when Ghdl_Rtik_Type_Array =>
+            Base_Array := To_Ghdl_Rtin_Type_Array_Acc(Rtii.Typ.Rti);
+            Layout_Addr := Rtii.Typ.Binding_Layout_Addr;
+         when Ghdl_Rtik_Subtype_Array =>
+            Base_Array := To_Ghdl_Rtin_Type_Array_Acc(
+               To_Ghdl_Rtin_Subtype_Composite_Acc(Rtii.Typ.Rti).Basetype);
+            Layout_Addr := Rtii.Typ.Layout_Addr;
+         when others =>
+            Internal_Error("get_array_child");
+      end case;
```

Comment:
You should create a function to get layout_addr and base_array from the type rtii.